### ISSUE TITLE
Clip gradient

### ISSUE THREAD

**Issue by haonan3** [DPBayes/ADADP#1]
It seems like, the norm computed in the following line is not the L2-norm of each layer's gradient, but the square root of the sum of gradient's squares of the different layers. Does it follow the algorithm mentioned in the paper "Learning rate adaptation for differentially private stochastic gradient descent."? I am a little bit confused because I think the norm of the gradient of the different layer should be taken separately.

https://github.com/DPBayes/ADADP/blob/866220450ad1f44384dc051486fb0375647a2f0d/MNIST_tests/px_expander.py#L32

**Comment by koskeant** [comment]
Thanks for your comment! This does follow the algorithm depicted in the paper. By summing the squared elements layer by layer, we get the total sum of squared elements of the gradient. After that, by taking the square root we get the L2-norm of the total gradient. This corresponds to computing the L2-norm of the gradient of each layer, then squaring these norms, summing and taking a square root.

**Comment by haonan3** [comment]
Thanks for your reply! The implementation is correct. It's my bad because I was thinking maybe we should give different clip-C for different layers.
